### DISCUSSION TITLE
Co-op in Private Parties (and Server Builds)

### DISCUSSION THREAD

**Discussion by Dliwk** [efroemling/ballistica#278] [Ideas]
However, I don't know how to change the levels... simply repeat selected one?.. Make possible to create co-op playlists?

In general, I would like to have the last option: choose any unlocked levels or some added with mods.

**Comment by efroemling** [comment]
I agree this would be very nice to have. I've just avoided it because the code to do so right now would be a bit ugly.  But if anyone wants to tackle it, the steps involved would basically be:

- Add a way to specify a co-op level in bacommon.servermanager.ServerConfig and wire up ba._servermode to honor that option and kick off a co-op session. Private parties simply set up server configs and start up servers so this would apply to them too.
- Wire up co-op sessions to simply repeat the same level when anyone presses a button or something at the score screen. Currently they wait for a UI button press on the host device but that will never happen because it is the server in this case. We would have to do something fancier if we wanted clients to be able to select different levels but I think repeating would be fine for now. They could restart the server when they wanted to switch levels.
- Update the playlist selection UI in the Private Server section to allow selecting unlocked Co-Op levels and not just Teams/FFA playlists. Note that this would need to be limited to built-in levels currently because the public host machines don't have any mods installed. I hope to allow user code on servers eventually but I need to do lots of hardening first due to the obvious security issues.

So if someone could get things to that point I'd be happy to wire up the final bits (communicating the co-op level selection to the master server so it can properly kick off the game).

I can try to get to this myself sometime too, but I'm currently focused on other 1.7 related things.

**Reply by Dliwk** [reply]
I've been poking around in this direction, but I'm a little confused:
- CoopScoreScreen's `on_player_join` / CoopSession's `on_player_leave` is never called in headless builds.. I've put print in them and i've seen it only in gui build :thinking:  .
- Is the end of session when all players left set *only* in `_coopsession.CoopSession.on_player_left`?
- Well, a few more strange things

**Reply by efroemling** [reply]
> * CoopScoreScreen's `on_player_join` / CoopSession's `on_player_leave` is never called in headless builds.. I've put print in them and i've seen it only in gui build 🤔  .

Hmm; I don't recall any logic that would cause these to only be run in gui builds 🤔. But then again I've never tried running co-op games under a headless build so perhaps there's something I'm not thinking of. I'd suggest throwing prints() everywhere to track down why this might be happening..

> * Is the end of session when all players left set _only_ in `_coopsession.CoopSession.on_player_left`?

Yeah that should cover the case when all players leave. Other ways of exiting such as the main menu will go through ba.app.return_to_main_menu_session_gracefully() I think...

**Reply by Dliwk** [reply]
>CoopScoreScreen's on_player_join / CoopSession's on_player_leave is never called in headless builds.. I've put print in them and i've seen it only in gui build :thinking:.

300 iq issue - i've just never seen output when running via `make prefab-server-debug` (some crazy output buffer stuff or what). Pretty strange that I've seen it when `make prefab-gui-debug`...

**Comment by MatiasMGz** [comment]
Any updates on this? Looking for this!
